pkg/jobWorker/job.go

```diff
@@ -0,0 +1,413 @@
+				}
+			}
+			cleanCGroup <- true
+		}
+	}()
+
+	err := ns.CreateCGroup(job.getCGroupName())
+	if err != nil {
+		cleanCGroup <- true
+		defer waitingCleanCGroupToCompleted()
+		return fmt.Errorf("error creating cgroup: %w", err)
+	}
+
+	if err = ns.AddResourceControl(job.getCGroupName(), ns.CpuWeightFile, strconv.Itoa(int(job.config.CPU*100))); err != nil {
+		cleanCGroup <- true
+		defer waitingCleanCGroupToCompleted()
+		return fmt.Errorf("could not add resources into controller:%s, %v", ns.CpuWeightFile, err)
+	}
+	if err = ns.AddResourceControl(job.getCGroupName(), ns.MemoryHighFile, strconv.FormatInt(job.config.MemBytes, 10)); err != nil {
+		cleanCGroup <- true
+		defer waitingCleanCGroupToCompleted()
+		return fmt.Errorf("could not add resources into controller:%s, %v", ns.MemoryHighFile, err)
+	}
+	//if err = ns.AddResourceControl(job.getCGroupName(), ns.IoWeightFile, strconv.FormatInt(job.config.IOBytesPerSecond, 10)); err != nil {
+	//	cleanCGroup <- 1
+	//	defer waitingCleanCGroupToCompleted()
+	//	return fmt.Errorf("could not add resources into controller:%s, %v", ns.IoWeightFile, err)
+	//}
+
+	//provide the file descriptor to cmd.Run so that it can add the new PID to the control group
+	if err = ns.AddProcess(job.getCGroupName(), cmd); err != nil {
+		cleanCGroup <- true
+		defer waitingCleanCGroupToCompleted()
+		return fmt.Errorf("error AddProcess /proc - %w\n", err)
+	}
+
+	unmount := make(chan bool)
+	waitingUnMountToCompleted := func() {
+		_ = <-unmount
+	}
+	go func() {
+		time.Sleep(100 * time.Millisecond)
+		select {
+		case <-unmount:
+			{
+				if err := ns.UnmountProc(); err != nil {
+					job.exitReason = errors.Join(job.exitReason, fmt.Errorf("error unmounting /proc - %w\n", err))
+				}
+				unmount <- true // unmouned
+			}
+		}
+	}()
+
+	if err = ns.MountProc(); err != nil {
+		cleanCGroup <- true
+		defer waitingUnMountToCompleted()
+		return fmt.Errorf("Error mounting /proc - %w\n", err)
+	}
+
+	log.Printf("starting job:%s, cmd:%s", job, cmd.String())
+	if err = cmd.Start(); err != nil {
+		cleanCGroup <- true
+		defer waitingCleanCGroupToCompleted()
+		unmount <- true
+		defer waitingUnMountToCompleted()
+		return fmt.Errorf("error starting command: %w", err)
+	}
+	job.cmd = cmd
+	job.isStarted = true
+
+	// run the command in a Goroutine so that Start can return immediately
+	go func() {
+		// Use cmd.Process.Wait() instead of cmd.Wait() since cmd.Wait() is not thread safe
+		// and we do not want to hold the mutex while waiting for the process to exit.
+		// So instead we use cmd.Process.Wait() and store the result in j.processState to mimic what
+		// cmd.Wait would do.
+		// This prevents concurrency issues when a user calls Start(), the command quickly exits (updating the
+		// process state), and the user invokes Status().
+		processState, err := job.cmd.Process.Wait()
+		job.processState = processState
+		job.exitCode = processState.ExitCode()
+		job.exitReason = job.cmd.Err
```

Comment:
Race condition those writes are unprotected by any lock  and it the a user calls job.Status will cause data race.

Reply:
Resolved: Moved after Lock to prevent race conditions.